File: lib/features/support/support_repository.dart
import 'package:dio/dio.dart';

import 'package:aroosi_flutter/core/api_client.dart';

/// Repository to submit support/contact requests.
/// Tolerant to multiple backend endpoint paths and payload shapes
/// to mirror aroosi-mobile behavior.
class SupportRepository {
  SupportRepository({Dio? dio}) : _dio = dio ?? ApiClient.dio;
  final Dio _dio;

  /// Submit a contact/support request.
  /// Returns true if the server accepted the request (2xx).
  Future<bool> submitContact({
    required String message,
    String? email,
    String? subject,
    String? category,
    Map<String, dynamic>? metadata,
  }) async {
    // Build multiple payload variants to maximize compatibility across backends.
    final payloads = <Map<String, dynamic>>[
      {
        if (email != null && email.isNotEmpty) 'email': email,
        if (subject != null && subject.isNotEmpty) 'subject': subject,
        if (category != null && category.isNotEmpty) 'category': category,
        'message': message,
        if (metadata != null && metadata.isNotEmpty) 'metadata': metadata,
      },
      {
        if (email != null && email.isNotEmpty) 'from': email,
        if (subject != null && subject.isNotEmpty) 'title': subject,
        if (category != null && category.isNotEmpty) 'type': category,
        'body': message,
        if (metadata != null && metadata.isNotEmpty) 'meta': metadata,
      },
      {
        if (email != null && email.isNotEmpty) 'userEmail': email,
        if (subject != null && subject.isNotEmpty) 'topic': subject,
        if (category != null && category.isNotEmpty) 'topicCategory': category,
        'description': message,
        if (metadata != null && metadata.isNotEmpty) 'extra': metadata,
      },
    ];

    final endpoints = <String>[
      '/support/contact',
      '/support',
      '/help/contact',
      '/contact',
    ];

    for (final data in payloads) {
      for (final path in endpoints) {
        try {
          final res = await _dio.post(path, data: data);
          final code = res.statusCode ?? 200;
          if (code >= 200 && code < 300) {
            return true;
          }
        } on DioException catch (_) {
          // Try next
        } catch (_) {
          // Try next
        }
      }
    }
    return false;
  }
}

File: native_ios/Sources/Shared/Services/Repositories/SupportRepository.swift
import Foundation

@available(iOS 17.0.0, *)
public struct SupportContactRequest: Equatable {
    public enum Category: String, CaseIterable {
        case general = "General"
        case technical = "Technical"
        case safety = "Safety"
    }

    public var email: String?
    public var subject: String?
    public var category: Category
    public var message: String
    public var metadata: [String: String]?

    public init(email: String? = nil,
                subject: String? = nil,
                category: Category = .general,
                message: String,
                metadata: [String: String]? = nil) {
        self.email = email
        self.subject = subject
        self.category = category
        self.message = message
        self.metadata = metadata
    }
}

@available(iOS 17.0.0, *)
public protocol SupportRepository {
    func submitContact(_ request: SupportContactRequest) async throws -> Bool
}

#if os(iOS)
#if canImport(FirebaseFirestore)
@available(iOS 17.0.0, macOS 12.0, *)
public final class RemoteSupportRepository: SupportRepository {
    private let client: HTTPClientProtocol
    private let logger = Logger.shared

    public init(client: HTTPClientProtocol? = nil) {
        if let client {
            self.client = client
        } else {
            self.client = try! DefaultHTTPClient()
        }
    }

    public func submitContact(_ request: SupportContactRequest) async throws -> Bool {
        let payloads = buildPayloads(from: request)
        for payload in payloads {
            for path in endpoints {
                do {
                    let body = try JSONSerialization.data(withJSONObject: payload, options: [])
                    let httpRequest = HTTPRequest(
                        path: path,
                        method: .post,
                        headers: ["Content-Type": "application/json"],
                        body: body
                    )
                    let (_, response) = try await client.data(for: httpRequest)
                    if (200..<300).contains(response.statusCode) {
                        return true
                    }
                } catch {
                    if #available(iOS 15.0, macOS 12.0, *), let cancellationError = error as? CancellationError {
                        throw cancellationError
                    }
                    logger.error("Support request to \(path) failed: \(error.localizedDescription)")
                }
            }
        }
        return false
    }

    private var endpoints: [String] {
        [
            "/support/contact",
            "/support",
            "/help/contact",
            "/contact"
        ]
    }

    private func buildPayloads(from request: SupportContactRequest) -> [[String: Any]] {
        var payloads: [[String: Any]] = []
        let metadata = request.metadata ?? [:]

        payloads.append([
            "email": request.email ?? "",
            "subject": request.subject ?? "",
            "category": request.category.rawValue,
            "message": request.message,
            "metadata": metadata
        ].filteringEmptyStrings())

        payloads.append([
            "from": request.email ?? "",
            "title": request.subject ?? "",
            "type": request.category.rawValue,
            "body": request.message,
            "meta": metadata
        ].filteringEmptyStrings())

        payloads.append([
            "userEmail": request.email ?? "",
            "topic": request.subject ?? "",
            "topicCategory": request.category.rawValue,
            "description": request.message,
            "extra": metadata
        ].filteringEmptyStrings())

        return payloads
    }
}

private extension Dictionary where Key == String, Value == Any {
    func filteringEmptyStrings() -> [String: Any] {
        var result: [String: Any] = [:]
        for (key, value) in self {
            if let string = value as? String {
                let trimmed = string.trimmingCharacters(in: .whitespacesAndNewlines)
                if !trimmed.isEmpty {
                    result[key] = trimmed
                }
            } else if let dictionary = value as? [String: Any], !dictionary.isEmpty {
                result[key] = dictionary
            } else if let anyValue = value as? [String], !anyValue.isEmpty {
                result[key] = anyValue
            } else if !(value is String) {
                result[key] = value
            }
        }
        return result
    }
}
#else
@available(iOS 17.0.0, *)
public final class RemoteSupportRepository: SupportRepository {
    public init() {}
    public func submitContact(_ request: SupportContactRequest) async throws -> Bool { false }
}
#endif
#else
@available(iOS 17.0.0, *)
public final class RemoteSupportRepository: SupportRepository {
    public init() {}
    public func submitContact(_ request: SupportContactRequest) async throws -> Bool { false }
}
#endif
